realm/src/main/java/io/realm/RealmQuery.java

```diff
@@ -1403,16 +1416,93 @@ public long count() {
      *
      * @return A {@link io.realm.RealmResults} containing objects. If no objects match the condition,
      * a list with zero objects is returned.
-     * @see io.realm.RealmResults
      * @throws java.lang.RuntimeException Any other error
      */
     public RealmResults<E> findAll() {
         return new RealmResults<E>(realm, query.findAll(), clazz);
     }
+    
+    /**
+     * Similar to {@link #findAll()} but runs asynchronously from a worker thread.
+     *
+     * @return immediately an empty {@link RealmResults}. Users need to register a listener
+     * {@link io.realm.RealmResults#addChangeListener(RealmChangeListener)} to be notified
+     * when the query completes.
+     * @throws java.lang.RuntimeException Any other error
+     * @see io.realm.RealmResults
+     */
+    public RealmResults<E> findAllAsync() {
+        // use caller Realm Looper
+        final WeakReference<Handler> weakHandler = new WeakReference<Handler>(realm.handler);
+
+        // handover the query (to be used by a worker thread)
+        final long handoverQueryPointer = query.handoverQuery(realm.sharedGroupManager.getNativePointer());
+
+        // save query arguments (for future update)
+        argumentsHolder =  new ArgumentsHolder(ArgumentsHolder.TYPE_FIND_ALL);
```

Comment:
We might want to restrict user from calling `findAllAsync()` and other async find functions from the same RealmQuery instance.
eg.
```
query.findAllAsync();
// Before findAllAsync finished, used called
query.findAllAsyncSorted();
```

If the first find failed because of sharedgroup mismatching, the query would be redo by `argumentsHolder` which has a value `TYPE_FIND_ALL_SORTED`. It should not be the expected behavior.